UKHomeOffice/workflow-service
# Cop 11437: Refactor and replace 'get integrity lead' subprocess with delegates

## Description
Currently there is a process in the Manual Declarations BPMN which uses a groovy script written directly in the BPMN to get the integrity leads for the job holder
Java delegate classes are used as a replacement for groovy scripts and are written in the Springboot application
Groovy scripts written directly in the BPMN are more difficult to read, control and test than Java delegates
This update has replaced the integrity lead groovy script with a Java delegate as a replicate of the original groovy script
The code does the following:
1. Finds users who have the group or role _bf-integrity-lead_
2. Gets the current user's team including _branchId_, _directorateId_ and _departmentId_
3. Finds integrity leads whose team's branchId matches the user's branchId
4. Failing this, finds integrity leads whose team's directorateId matches the user's
5. Failing this, finds integrity leads whose team's departmentId matches the user's
6. Failing this, finds users who are in the OAD team
7. IntegrityLeads which the user wishes to exclude are then removed from the list
8. The emails of the integrity leads are then passed on as an execution variable

## Testing
Several unit tests have been written testing:
- The service interacting with the identity service works
- The matching works correctly by _Branch_, _Directorate_, and _Department_
- The data are correct on matching
- OAD's are returned if no other match

Test BPMN's have also been created which logs out the results of both the current and new methods of getting integrity lead emails and the results are the same. Steps to reproduce this are as follows (BPMN's supplied on request):
1. Pull this branch of the workflow service
2. Run the workflow service with the correct VM options to work locally
3. Run `docker-compose up` in terminal from the workflow service directory to start the containers
4. Acquire Bearer tokens for the ref-data service and the identity service
5. Add the Bearer tokens to the headers for the requests tasks to ref-data and identity service in the match-integrity-leads BPMN
6. Deploy both BPMN's
7. Start the mandatory-declarations BPMN process
9. Complete the BPMN process to ensure it gets to the match-integrity-leads sub process
10. Observe the running workflow service console output
11. Confirm that the old method and the new method outputs for integrity lead emails are the same:
10a. `old method matched ils = someswar.duddi@digital.homeoffice.gov.uk,test_user@digital.homeoffice.gov.uk`
10b. `new method matched ils = someswar.duddi@digital.homeoffice.gov.uk,test_user@digital.homeoffice.gov.uk`
